### PR TITLE
datapath: Use TUNNEL_MODE as indicator for tunnel mode

### LINKED ISSUE #16147: Egress for destination CIDR 0.0.0.0/0 blocks all network connections
## Bug report

**General Information**

- Cilium version (run `cilium version`)
- Kernel version (run `uname -a`)
- Orchestration system version in use (e.g. `kubectl version`, ...)
- Link to relevant artifacts (policies, deployments scripts, ...)
- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```

**How to reproduce the issue**

1. Use the following egress rule
```
apiVersion: cilium.io/v2alpha1
kind: CiliumEgressNATPolicy
metadata:
  name: egress-mailserver
spec:
  egress:
  - podSelector:
      matchLabels:
        # The following label selects default namespace
        io.kubernetes.pod.namespace: mailserver
    # Or use namespace label selector to select multiple namespaces
    # namespaceSelector:
    #  matchLabels:
    #    ns: default
  destinationCIDRs:
  - 0.0.0.0/0
  egressSourceIP: "51.195.x.x" 
```
2. curl http://checkip.amazonaws.com from a pod
 
Expected result: 51.195.x.x
Actual result: bash-5.0# curl http://checkip.amazonaws.com
curl: (6) Could not resolve host: checkip.amazonaws.com

### PR DESCRIPTION
**Pull request by anfernee**
datapath: Use TUNNEL_MODE as indicator for tunnel mode
    
    Originally ENCAP_IFINDEX is used as indicator of tunnel mode, which
    works well before egress gateway. In egress gateway's direct routing
    mode, we also have a tunnel for forwarding traffic, meaning that we
    also have ENCAP_IFINDEX. This will cause broken and unexpected data
    flow:
    
    - Pod to pod traffic could be forwarded to tunnel in direct routing mode
    - Some node to pod traffic is broken in direct routing mode because the
    return packet is masqueraded by node IP.
    
    As a fix, add a new macro TUNNEL_MODE as indicator of tunnel mode,
    either vxlan or geneve.

Fixes: #16147

cc @MasterZ40 @pchaigno

### CODE REVIEW

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
test-1.21-4.9

File: bpf/init.sh [hunk 1]
```diff
@@ -421,6 +421,11 @@ else
 	ip link del cilium_sit   2> /dev/null || true
 fi
 
+if [ "$MODE" = "tunnel" ]; then
+	sed -i '/^#.*TUNNEL_MODE.*$/d' $RUNDIR/globals/node_config.h
+	echo "#define TUNNEL_MODE 1" >> $RUNDIR/globals/node_config.h
+fi
```
Comment (pchaigno) [inline bpf/init.sh:427]: So if I understand correctly, `TUNNEL_MODE` will only be defined if we are running in tunneling mode and enabling the egress gateway in direct routing mode will define `ENCAP_IFINDEX` but not `TUNNEL_MODE`?

File: bpf/Makefile [hunk 2]
```diff
@@ -148,22 +148,22 @@ bpf_overlay.o: bpf_overlay.ll
...
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_HOST_FIREWALL: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_HOST_FIREWALL: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:
 
 MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1
```
Comment (pchaigno) [inline bpf/Makefile]: Shouldn't `TUNNEL_MODE` be added here and for other `MAX_*_OPTIONS`?

**Review by pchaigno — COMMENTED** [review]
How did you test this? Did you run it through the WIP end-to-end tests?

**Comment by pchaigno** [comment]
There's a legitimate failure in tests:
```
	 clang -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -I/cilium/bpf/include -I/cilium/bpf -D__NR_CPUS__=3 -O2 -g -target bpf -std=gnu89 -nostdinc -emit-llvm -Wall -Wextra -Werror -Wshadow -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-gnu-variable-sized-type-not-at-end -Wdeclaration-after-statement -c bpf_host.c -o bpf_host.ll
	 make: Leaving directory '/cilium/bpf'
	 
Stderr:
 	 bpf_host.c:648:24: error: use of undeclared identifier 'IPV4_ENCRYPT_IFACE'
	         __u32 tunnel_source = IPV4_ENCRYPT_IFACE;
	                               ^
	 1 error generated.
```

File: bpf/init.sh (hunk 1 above)
Comment (anfernee) [reply bpf/init.sh:427]: > So if I understand correctly, TUNNEL_MODE will only be defined if we are running in tunneling mode 

Yes, `MODE` is set to `tunnel` when `--tunnel!=disabled`. This change will define `TUNNEL_MODE` when it's true:

https://github.com/cilium/cilium/blob/e92dc6ac6b766e793091410d0cf58c61b01d424d/pkg/datapath/loader/base.go#L345-L347

> enabling the egress gateway in direct routing mode will define `ENCAP_IFINDEX` but not `TUNNEL_MODE`?

That's correct.

File: bpf/Makefile (hunk 2 above)
Comment (anfernee) [reply bpf/Makefile]: Good point. Tunnel mode makes bpf program more complex, so I think it should.

**Review by pchaigno — APPROVED** [review]

**Review by kkourt — APPROVED** [review]

**Comment by nathanjsweet** [comment]
test-me-please

**Comment by anfernee** [comment]
> There's a legitimate failure in tests:
> 
> ```
> 	 clang -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_IPSEC=1 -DIP_POOLS=1 -DENCAP_IFINDEX=1 -I/cilium/bpf/include -I/cilium/bpf -D__NR_CPUS__=3 -O2 -g -target bpf -std=gnu89 -nostdinc -emit-llvm -Wall -Wextra -Werror -Wshadow -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-gnu-variable-sized-type-not-at-end -Wdeclaration-after-statement -c bpf_host.c -o bpf_host.ll
> 	 make: Leaving directory '/cilium/bpf'
> 	 
> Stderr:
>  	 bpf_host.c:648:24: error: use of undeclared identifier 'IPV4_ENCRYPT_IFACE'
> 	         __u32 tunnel_source = IPV4_ENCRYPT_IFACE;
> 	                               ^
> 	 1 error generated.
> ```

The test seems to refer to a Makefile in `test-verifier` image that is not updated. Looking at ways to fix it.

**Comment by anfernee** [comment]
ci-aks

**Comment by anfernee** [comment]
> How did you test this? Did you run it through the WIP end-to-end tests?

We did k8s conformance test internally on a similar backported change in 1.9 branch. It fixed the issue. I tried the WIP e2e test, but it still failed. But I think it's a different issue.

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
The aks connectivity error seems to be legit. It's caused by the second change https://github.com/cilium/cilium/pull/16328/commits/78be5e14eb2942cf134b50476129ffc7e4b63ced (forgot to mention: this is a change we didn't include in the local conformance test). Let me exclude that change first, and try to understand why it causes issues.

**Comment by anfernee** [comment]
78be5e1 shouldn't fail the aks test, because the change is all hidden behind `ENABLE_EGRESS_GATEWAY`, which is not even enabled. Is there some kind of known flakiness for the test?

**Comment by pchaigno** [comment]
> [78be5e1](https://github.com/cilium/cilium/commit/78be5e14eb2942cf134b50476129ffc7e4b63ced) shouldn't fail the aks test, because the change is all hidden behind `ENABLE_EGRESS_GATEWAY`, which is not even enabled. Is there some kind of known flakiness for the test?

Yes, that test is very flaky at the moment. We can ignore it here.

**Review by pchaigno — APPROVED** [review]

**Pull request merged**